//    Copyright 2019 Google Inc.
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//        https://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.

syntax = "proto3";

package google.fhir.r4.core;

import "google/protobuf/any.proto";
import "proto/annotations.proto";
import "proto/r4/core/codes.proto";
import "proto/r4/core/datatypes.proto";

option java_multiple_files = true;
option java_package = "com.google.fhir.r4.core";
option (.google.fhir.proto.fhir_version) = R4;

// Auto-generated from StructureDefinition for Patient, last updated
// 2018-12-27T22:37:54.724+11:00. Information about an individual or animal
// receiving health care services. See
// http://hl7.org/fhir/StructureDefinition/Patient
message Patient {
  option (.google.fhir.proto.structure_definition_kind) = KIND_RESOURCE;
  option (.google.fhir.proto.fhir_structure_definition_url) =
      "http://hl7.org/fhir/StructureDefinition/Patient";

  // Logical id of this artifact
  Id id = 1;

  // Metadata about the resource
  Meta meta = 2;

  // A set of rules under which this content was created
  Uri implicit_rules = 3;

  // Language of the resource content
  Code language = 4;

  // Text summary of the resource, for human interpretation
  Narrative text = 5;

  // Contained, inline Resources
  repeated protobuf.Any contained = 6;

  // Field 7 reserved for strongly-typed ContainedResource for id:
  // Patient.contained
  reserved 7;

  // Additional content defined by implementations
  repeated Extension extension = 8;

  // Extensions that cannot be ignored
  repeated Extension modifier_extension = 9;

  // An identifier for this patient
  repeated Identifier identifier = 10;

  // Whether this patient's record is in active use
  Boolean active = 11;

  // A name associated with the patient
  repeated HumanName name = 12;

  // A contact detail for the individual
  repeated ContactPoint telecom = 13;

  // male | female | other | unknown
  message GenderCode {
    option (.google.fhir.proto.structure_definition_kind) = KIND_PRIMITIVE_TYPE;
    option (.google.fhir.proto.fhir_valueset_url) =
        "http://hl7.org/fhir/ValueSet/administrative-gender";
    option (.google.fhir.proto.fhir_profile_base) =
        "http://hl7.org/fhir/StructureDefinition/code";

    AdministrativeGenderCode.Value value = 1;

    String id = 2;

    repeated Extension extension = 3;
  }
  GenderCode gender = 14;

  // The date of birth for the individual
  Date birth_date = 15;

  // Indicates if the individual is deceased or not
  message DeceasedX {
    option (.google.fhir.proto.is_choice_type) = true;

    oneof choice {
      Boolean boolean = 1;
      DateTime date_time = 2;
    }
  }
  DeceasedX deceased = 16;

  // An address for the individual
  repeated Address address = 17;

  // Marital (civil) status of a patient
  CodeableConcept marital_status = 18;

  // Whether patient is part of a multiple birth
  message MultipleBirthX {
    option (.google.fhir.proto.is_choice_type) = true;

    oneof choice {
      Boolean boolean = 1;
      Integer integer = 2;
    }
  }
  MultipleBirthX multiple_birth = 19;

  // Image of the patient
  repeated Attachment photo = 20;

  // A contact party (e.g. guardian, partner, friend) for the patient
  message Contact {
    option (.google.fhir.proto.fhir_path_message_constraint) =
        "name.exists() or telecom.exists() or address.exists() or organization.exists()";

    // Unique id for inter-element referencing
    String id = 1;

    // Additional content defined by implementations
    repeated Extension extension = 2;

    // Extensions that cannot be ignored even if unrecognized
    repeated Extension modifier_extension = 3;

    // The kind of relationship
    repeated CodeableConcept relationship = 4;

    // A name associated with the contact person
    HumanName name = 5;

    // A contact detail for the person
    repeated ContactPoint telecom = 6;

    // Address for the contact person
    Address address = 7;

    // male | female | other | unknown
    message GenderCode {
      option (.google.fhir.proto.structure_definition_kind) =
          KIND_PRIMITIVE_TYPE;
      option (.google.fhir.proto.fhir_valueset_url) =
          "http://hl7.org/fhir/ValueSet/administrative-gender";
      option (.google.fhir.proto.fhir_profile_base) =
          "http://hl7.org/fhir/StructureDefinition/code";

      AdministrativeGenderCode.Value value = 1;

      String id = 2;

      repeated Extension extension = 3;
    }
    GenderCode gender = 8;

    // Organization that is associated with the contact
    Reference organization = 9
        [(.google.fhir.proto.valid_reference_type) = "Organization"];

    // The period during which this contact person or organization is valid to
    // be contacted relating to this patient
    Period period = 10;
  }
  repeated Contact contact = 21;

  // A language which may be used to communicate with the patient about his or
  // her health
  message Communication {
    // Unique id for inter-element referencing
    String id = 1;

    // Additional content defined by implementations
    repeated Extension extension = 2;

    // Extensions that cannot be ignored even if unrecognized
    repeated Extension modifier_extension = 3;

    // The language which can be used to communicate with the patient about his
    // or her health
    CodeableConcept language = 4
        [(.google.fhir.proto.validation_requirement) = REQUIRED_BY_FHIR];

    // Language preference indicator
    Boolean preferred = 5;
  }
  repeated Communication communication = 22;

  // Patient's nominated primary care provider
  repeated Reference general_practitioner = 23 [
    (.google.fhir.proto.valid_reference_type) = "Organization",
    (.google.fhir.proto.valid_reference_type) = "Practitioner",
    (.google.fhir.proto.valid_reference_type) = "PractitionerRole"
  ];

  // Organization that is the custodian of the patient record
  Reference managing_organization = 24
      [(.google.fhir.proto.valid_reference_type) = "Organization"];

  // Link to another patient resource that concerns the same actual person
  message Link {
    // Unique id for inter-element referencing
    String id = 1;

    // Additional content defined by implementations
    repeated Extension extension = 2;

    // Extensions that cannot be ignored even if unrecognized
    repeated Extension modifier_extension = 3;

    // The other patient or related person resource that the link refers to
    Reference other = 4 [
      (.google.fhir.proto.validation_requirement) = REQUIRED_BY_FHIR,
      (.google.fhir.proto.valid_reference_type) = "Patient",
      (.google.fhir.proto.valid_reference_type) = "RelatedPerson"
    ];

    // replaced-by | replaces | refer | seealso
    message TypeCode {
      option (.google.fhir.proto.structure_definition_kind) =
          KIND_PRIMITIVE_TYPE;
      option (.google.fhir.proto.fhir_valueset_url) =
          "http://hl7.org/fhir/ValueSet/link-type";
      option (.google.fhir.proto.fhir_profile_base) =
          "http://hl7.org/fhir/StructureDefinition/code";

      LinkTypeCode.Value value = 1;

      String id = 2;

      repeated Extension extension = 3;
    }
    TypeCode type = 5
        [(.google.fhir.proto.validation_requirement) = REQUIRED_BY_FHIR];
  }
  repeated Link link = 25;
}
